routes/common/vm_helper.rb

```diff
@@ -26,80 +26,56 @@ def list
-        parsed_size = Validation.validate_vm_size(request_body_params["size"], only_visible: true)
-      end
+    # Same as above, moved the size validation here to not allow users to
+    # pass gpu instance while creating a VM.
+    if request_body_params["size"]
+      parsed_size = Validation.validate_vm_size(request_body_params["size"], only_visible: true)
+    end
 
-      if request_body_params["private_subnet_id"]
-        ps = PrivateSubnet.from_ubid(request_body_params["private_subnet_id"])
-        unless ps && ps.location == @location
-          fail Validation::ValidationFailed.new({private_subnet_id: "Private subnet with the given id \"#{request_body_params["private_subnet_id"]}\" is not found in the location \"#{LocationNameConverter.to_display_name(@location)}\""})
-        end
-        Authorization.authorize(@user.id, "PrivateSubnet:view", ps.id)
-        request_body_params["private_subnet_id"] = ps.id
-      end
+    if request_body_params["storage_size"]
+      storage_size = Validation.validate_vm_storage_size(request_body_params["size"], request_body_params["storage_size"])
+      request_body_params["storage_volumes"] = [{size_gib: storage_size, encrypted: true}]
+      request_body_params.delete("storage_size")
+    end
 
-      if request_body_params["storage_size"]
-        storage_size = Validation.validate_vm_storage_size(request_body_params["size"], request_body_params["storage_size"])
-        request_body_params["storage_volumes"] = [{size_gib: storage_size, encrypted: true}]
-        request_body_params.delete("storage_size")
+    if request_body_params["private_subnet_id"] && request_body_params["private_subnet_id"] != ""
+      ps = PrivateSubnet.from_ubid(request_body_params["private_subnet_id"])
+      if !ps || ps.location != @location
+        fail Validation::ValidationFailed.new({private_subnet_id: "Private subnet with the given id \"#{request_body_params["private_subnet_id"]}\" is not found in the location \"#{LocationNameConverter.to_display_name(@location)}\""})
       end
+      Authorization.authorize(@user.id, "PrivateSubnet:view", ps.id)
+    end
+    request_body_params["private_subnet_id"] = ps&.id
 
-      requested_vm_core_count = parsed_size.nil? ? 1 : parsed_size.vcpu / 2
-      Validation.validate_core_quota(project, "VmCores", requested_vm_core_count)
+    requested_vm_core_count = parsed_size.nil? ? 1 : parsed_size.vcpu / 2
+    Validation.validate_core_quota(project, "VmCores", requested_vm_core_count)
 
-      st = Prog::Vm::Nexus.assemble(
-        request_body_params["public_key"],
-        project.id,
-        name: name,
-        location: @location,
-        **request_body_params.except(*required_parameters).transform_keys(&:to_sym)
-      )
+    st = Prog::Vm::Nexus.assemble(
+      request_body_params["public_key"],
+      project.id,
+      name: name,
+      location: @location,
+      **request_body_params.except(*required_parameters).transform_keys(&:to_sym)
```

Comment:
wew, these kinds of splats are quite dangerous. I realize there was pre-existence, so the onus is not on this refactoring, but on balance, I'd rather define and pack a struct and pass it, or something like that, given you already have a enumeration of field names.

Reply:
@geemus @enescakir got any thoughts or tips on this one?

Reply:
This also caught my eye during refactoring. I left it as is because at line 33, we call `Validation.validate_request_body(params, required_parameters, allowed_optional_parameters)`, which ensures that we only accept predefined set of parameters.

Of course, future refactorings could endanger this. I think we can directly use `allowed_optional_parameters` to generate an approved list and use splat operator on that list.

Reply:
I'll fix this in a separate issue. It is also good opportunity to use new `security` tag.